### PR TITLE
Choice instances for cross sdk

### PR DESCRIPTION
**Pull request by nickchapman-da**
`Upgrade.hs`
- Generate instances for template choices
- generate new 5-element format for `external` strings

`LFConversion.hs`
- (aside: fix text in existing "is not supported" messages)
- support `external` strings: `[pkgId, modStr, templName, choiceName, method]`

`GenerateSimpleDalf.hs`
- fix generation of `Choice` to be well-formed:
    - (1) Name of choice matches name of datatype
    - (2) datatype is a record (not a variant!)

`IntegrationTests.hs`
- refactor to split `packagingTests` into `packagingTests` / `crossSdkTests` / `migrationTests`
- Extend `crossSdkTests` to test choice-exercise:
    `submit bob $ exercise coid1 Module.Choice with choiceArg`

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs [hunk 1]
```diff
@@ -152,6 +153,21 @@ templateInstances env externPkgId =
   where
     mod = envMod env
 
+choiceInstances :: Env -> LF.PackageId -> [HsDecl GhcPs]
+choiceInstances env externPkgId =
+    [ generateChoiceInstance env externPkgId templateDT choice
+    | template <- NM.elems $ LF.moduleTemplates mod
+    , choice <- NM.elems $ LF.tplChoices template
+    , not (specialArchiveChoice choice)
```
Comment (cocreature) [inline compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: Why do we filter out `Archive`? Yes, we have an `archive` method in the `Template` typeclass but we also generate `Choice` instances for it and that matters if you want to use code that is polymorphic in a `Choice` constraint so just calling `archive` isn’t a workaround.

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs [hunk 2]
```diff
@@ -152,6 +153,21 @@ templateInstances env externPkgId =
...
+    , Just templateDT <-
```
Comment (cocreature) [inline compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: Currently we will silently discard all choices if this lookup fails. Afaict if this lookup fails it’s a bug somewhere so crashing seems like a better option than silently discarding the choices.

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs [hunk 3]
```diff
@@ -234,7 +250,7 @@ generateTemplateInstance ::
     -> [(LF.TypeVarName, LF.Kind)]
     -> LF.PackageId
     -> HsDecl GhcPs
-generateTemplateInstance env typeCon typeParams externPkgId =
+generateTemplateInstance env dataTypeCon dataParams externPkgId =
```
Comment (cocreature) [inline compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: The renaming seems slightly confusing in `data X = Y`, `X` is a type constructor and `Y` is a data constructor. I’ve never heard the term datatype constructor. As for `dataParams` this sounds like it might refer to the fields in a data constructor.

Leaving the naming aside, is there ever a case where `dataParams` is non-empty at the moment? Given that we monomorphize types in DAML-LF, I’m having trouble coming up with a way to hit this. If not, it would be good to document this in a comment (this will most likely change in DAML-LF soonish so no need to remove it completely).

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs [hunk 4]
```diff
@@ -382,7 +382,7 @@ convertGenericTemplate env x
                                             (ETmLam (mkVar "key", keyType) $ ERecCon anyContractKeyTy [(anyContractKeyField, EToAny keyType $ EVar $ mkVar "key")]))
                                   else EBuiltin BEError `ETyApp`
                                        TForall (mkTypeVar "proxy", KArrow KStar KStar) (TApp (TVar $ mkTypeVar "proxy") polyType :-> keyType :-> typeConAppToType anyContractKeyTy) `ETmApp`
-                                       EBuiltin (BEText "toAnyChoice is not supported in this DAML-LF version")
+                                       EBuiltin (BEText "toAnyContractKey is not supported in this DAML-LF version")
```
Comment (cocreature) [inline compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs]: Good catch, thanks!

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs [hunk 5]
```diff
@@ -283,86 +307,156 @@ generateTemplateInstance env typeCon typeParams externPkgId =
...
+      noLoc $ convType env lfTemplateType
+
+    arg2 :: LHsType GhcPs =
+      noLoc $ convType env lfChoiceType
+
+    arg3 :: LHsType GhcPs =
+      noLoc $ convType env lfChoiceReturnType
+
+    moduleNameStr = T.unpack $ LF.moduleNameString $ LF.moduleName $ envMod env
+    moduleName0 =
```
Comment (cocreature) [inline compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: Isn’t this just `LF.moduleName $ envMod mod`? It looks like you are first intercalating dots via `moduleNameString`, unpacking only to then split on dots and repack.

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs [hunk 6]
```diff
@@ -283,86 +307,156 @@ generateTemplateInstance env typeCon typeParams externPkgId =
...
+        LF.ModuleName $
+        map T.pack $
+        splitOn "." moduleNameStr
+
+    lfTemplateType :: LF.Type =
+        LF.mkTApps
+          (LF.TCon (LF.Qualified LF.PRSelf moduleName0 dataTypeCon))
+          (map (LF.TVar . fst) dataParams)
+
+    lfChoiceType :: LF.Type =
```
Comment (cocreature) [inline compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: This isn’t quite right:
1. The choice type does not need to be defined in the same module as the template. This matters both for `Archive` (is that why you filtered it out above?) and for generic templates where the monomorphized version will live in the module where you create the instance while the choice type will live in the module where you define the generic template.
2. In DAML-LF the name of the choice type constructor must not match the choice name. There is no good way to represent this in DAML and we never generate such code so I would suggest to error out in that case.

So in summary:
* Validate that the choice name and the type constructor name of the choice type match
* Use the choice argument type instead of inventing the choice type based on the choice name.

File: compiler/damlc/tests/src/DA/Test/GenerateSimpleDalf.hs [hunk 7]
```diff
@@ -47,7 +47,7 @@ main = do
             , dataTypeCon = TypeConName ["Choice"]
             , dataSerializable = IsSerializable True
             , dataParams = []
-            , dataCons = DataVariant [(VariantConName "Choice", TUnit)]
+            , dataCons = DataRecord [ (FieldName "choiceArg", TUnit) ]
```
Comment (cocreature) [inline compiler/damlc/tests/src/DA/Test/GenerateSimpleDalf.hs]: Why can’t we support variants here? I know that it’s not possible to generate such code from DAML since we always build up a record (unless you manually write `Choice` instances) but it seems fairly easy to support this. in fact, it probably works out of the box unless you validate that the type is a record which I think you don’t do at the moment.

If we don’t support it as part of this PR, that’s fine as well but in that case I would like to keep the test case as an expected failure.

File: compiler/damlc/tests/src/DA/Test/GenerateSimpleDalf.hs [hunk 8]
```diff
@@ -58,7 +58,7 @@ main = do
             }
     let chc = TemplateChoice
             { chcLocation = Nothing
-            , chcName = ChoiceName "NotChoice"
+            , chcName = ChoiceName "Choice"
```
Comment (cocreature) [inline compiler/damlc/tests/src/DA/Test/GenerateSimpleDalf.hs]: This hits exactly the case that I mentioned above where the choice name and type constructor name don’t match. Let’s please keep this and test that we emit a somewhat usable error message instead of simply removing it.

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs (hunk 6 above)
Comment (nickchapman-da) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: Yes that's why I filtered `Archive`.
I was assuming that for a well-formed DAML the names must match.
If not, how do we discover from a `TemplateChoice` the corresponding `TypeConName`?

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs (hunk 3 above)
Comment (nickchapman-da) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: The reason for the name change is to match the names used by the only caller of the function.
And at this point, the names are bound by the `LF.DefDataType{..}`
`dataTypeCon` and `dataParams` being the field names of `DefDataType`.

Regarding whether `dataParams` can be non-empty: I'm not sure. This code was pre-existing. So I assume Robin passed this because it could be non-empty.  @robin-da ?

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs (hunk 6 above)
Comment (cocreature) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: In DAML the names match. In DAML-LF the definition of a template choice includes the name as well as the type as separate fields and no relationship between the two is required.

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs (hunk 3 above)
Comment (cocreature) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: It makes some sense in the datatype definition due to Haskell’s issues with colliding record field names. I don’t think it makes any sense here.

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs (hunk 2 above)
Comment (nickchapman-da) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: My defence is I was matching the style of existing code in `templateInstances`, so I assumed it was possible for this lookup to fail. But it does seem unlikely. @robin-da ?

File: compiler/damlc/tests/src/DA/Test/GenerateSimpleDalf.hs (hunk 7 above)
Comment (nickchapman-da) [reply compiler/damlc/tests/src/DA/Test/GenerateSimpleDalf.hs]: The reason to change to a record type was because I wanted to test a cross-sdk exercise using DAML. I'll make the generated DALF have 2 choices: One with a record; one with a variant.

File: compiler/damlc/tests/src/DA/Test/GenerateSimpleDalf.hs (hunk 8 above)
Comment (nickchapman-da) [reply compiler/damlc/tests/src/DA/Test/GenerateSimpleDalf.hs]: If the related type of a choice can be anything, and we support that, the why would this case be an error?

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs (hunk 6 above)
Comment (nickchapman-da) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: The type being the 2nd element of the pair `chcArgBinder`, right?

File: compiler/damlc/tests/src/DA/Test/GenerateSimpleDalf.hs (hunk 7 above)
Comment (cocreature) [reply compiler/damlc/tests/src/DA/Test/GenerateSimpleDalf.hs]: Sounds good :+1:

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs (hunk 6 above)
Comment (cocreature) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: There are a couple of different restrictions at different levels so let me try to make them explicit (also responding to the comment below on why we probably don’t want to support mismatches between the choice name and the choice type constructor name).

1. If you define a template in DAML, we will generate a record type for each choice with the same name as the choice. There are two somewhat special cases here:
    1. The `Archive` record is defined in `daml-stdlib` so there is only one for all templates.
    2. For generic templates, we only defined one choice record that might have type parameters instead of defining one per instance.

 However, even for these cases the choice name is the same as the type constructor name.
2. The `Choice` typeclass in DAML requires the pair `(templateType, choiceArgType)` to uniquely identify a choice on a given template. In particular, that means that we cannot represent two choices with the same argument type in DAML. It does _not_ require that `choiceArgType` is a record and supporting variants should work out nicely and not be confusing. The `Choice` typeclass does also _not_ require that the type constructor name  and the choice name match. However, if that is not the case things get really confusing. For an extreme example consider a choice `Foo` which takes a type called `Bar` as the argument and a choice called `Bar` which takes a type called `Foo` as the argument. To call the `Foo` choice you would now do something like `exercise cid Bar` in DAML which at least to me isn’t intuitive at all. Furthermore, while you could support this case it falls apart if you have a template with two choices (with different names) that take the same choice argument type. Therefore, the most sensible option (at least initially) seems to me to reject choices that have different choice and type constructor names.
3. In DAML-LF the only restriction is that the choice names must unique per template. It is possible to use the same choice argument type for multiple choices, it is possible to use non-record types (e.g., even `()` which we should also reject). As mentioned above the `Choice` typeclass above does not allow us to express this in full generality so we necessarily have to reject some cases and we probably want to reject slightly more than strictly necessary to provide a reasonable UX.

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs (hunk 2 above)
Comment (robin-da) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: it's fine to drop the choice if it's not present in the dalf, but here you will drop it even if it's present but a later condition will fail. That's different from `templateInstances`, I think this case should be a crash.

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs (hunk 1 above)
Comment (nickchapman-da) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: no longer filter Archive

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs (hunk 2 above)
Comment (nickchapman-da) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: change to crash if cant find the template datatype def

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs (hunk 3 above)
Comment (nickchapman-da) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: reverted name change

Comment (nickchapman-da) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: the question about if `dataParams` can be non-empty is still open

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs (hunk 6 above)
Comment (nickchapman-da) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: Fixed code to no longer assume there is a datatype named the same as the choice

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs (hunk 5 above)
Comment (nickchapman-da) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: simplified to avoid this

File: compiler/damlc/tests/src/DA/Test/GenerateSimpleDalf.hs (hunk 8 above)
Comment (nickchapman-da) [reply compiler/damlc/tests/src/DA/Test/GenerateSimpleDalf.hs]: reinstated to name the choice as "NotChoice" in the test (which works now because the code no longer assumes the choice datatype is named the same as the choice)

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs [hunk 9]
```diff
@@ -112,8 +112,7 @@ generateTemplateInstanceModule env externPkgId
               map (showSDocForUser fakeDynFlags alwaysQualify . ppr) instances)
     | otherwise = Nothing
   where
-    instances = templInstances
-    templInstances = templateInstances env externPkgId
+    instances = templateInstances env externPkgId ++ choiceInstances env externPkgId
```
Comment (cocreature) [inline compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: btw, in case you’re still looking for what to work on next, we also need `TemplateKey` instances.

File: compiler/damlc/tests/src/DA/Test/GenerateSimpleDalf.hs (hunk 7 above)
Comment (cocreature) [reply compiler/damlc/tests/src/DA/Test/GenerateSimpleDalf.hs]: Looks like you didn’t do this? We still seem to only have one choice for a `Record`.

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs (hunk 3 above)
Comment (cocreature) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: I’m fine leaving it for now and we can sort it out later.

File: compiler/damlc/tests/src/DA/Test/GenerateSimpleDalf.hs (hunk 8 above)
Comment (cocreature) [reply compiler/damlc/tests/src/DA/Test/GenerateSimpleDalf.hs]: As I said elsewhere, I would like us to emit an error if the choice name and the type constructor name don’t match. Having  `exercise cid Choice` call a choice called `NotChoice` is really confusing.

File: compiler/damlc/tests/src/DA/Test/GenerateSimpleDalf.hs (hunk 7 above)
Comment (nickchapman-da) [reply compiler/damlc/tests/src/DA/Test/GenerateSimpleDalf.hs]: done now. 
But I'm not really convinced of the utility of creating DALF for testing which is not of the form which would result from the daml compiler.

File: compiler/damlc/tests/src/DA/Test/GenerateSimpleDalf.hs (hunk 8 above)
Comment (nickchapman-da) [reply compiler/damlc/tests/src/DA/Test/GenerateSimpleDalf.hs]: I'm confused as to where you think this error check should go.

First, we should decide it this name-match really is a well-formed condition of DALF, or not.

If it isn't, then a mismatched name is confusing I agree, but not an error to be detected, and especially not when importing existing DALF via cross-sdk.

If it is an explicit well-formed condition on DALF, then perhaps the compiler should assert this condition just like it does for the well-typed condition.  And probably we shouldn't generate testing DALF which is not well-formed.

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs (hunk 9 above)
Comment (nickchapman-da) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: I'll discuss with Robin who is best placed to work on these.

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs [hunk 10]
```diff
@@ -1660,8 +1660,34 @@ convertExternal env stdlibRef primId lfType
...
+    , Just TemplateChoice {chcSelfBinder,chcArgBinder} <- NM.lookup choice tplChoices = do
+        let pkgRef = PRImport $ PackageId $ T.pack pkgId
+        let mod = ModuleName $ map T.pack $ splitOn "." modStr
+        let qualify tconName = Qualified { qualPackage = pkgRef , qualModule = mod , qualObject = tconName}
+        let templateDataType = TCon . qualify
+        let (choiceArg, choiceArgType) = chcArgBinder
+        case method of
+          "exercise" -> do
+            ETmLam (chcSelfBinder, TApp (TBuiltin BTContractId) (templateDataType tplTypeCon)) $
+              ETmLam (choiceArg, choiceArgType) $
```
Comment (robin-da) [inline compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs]: you can just use the `chcArgBinder` variable here.

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs [hunk 11]
```diff
@@ -1660,8 +1660,34 @@ convertExternal env stdlibRef primId lfType
     | [pkgId, modStr, templName, method] <- splitOn ":" primId
     , Nothing <- lookup pkgId modStr templName =
         error $ "convertExternal: external template not found " <> primId
+    | [pkgId, modStr, templName, choiceName, method] <- splitOn ":" primId
+    , Just LF.Template {tplTypeCon,tplChoices} <- lookup pkgId modStr templName
+    , choice <- ChoiceName (T.pack choiceName)
```
Comment (robin-da) [inline compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs]: This is just a let binding, right?

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs [hunk 12]
```diff
@@ -1660,8 +1660,34 @@ convertExternal env stdlibRef primId lfType
...
+        let qualify tconName = Qualified { qualPackage = pkgRef , qualModule = mod , qualObject = tconName}
+        let templateDataType = TCon . qualify
+        let (choiceArg, choiceArgType) = chcArgBinder
+        case method of
+          "exercise" -> do
+            ETmLam (chcSelfBinder, TApp (TBuiltin BTContractId) (templateDataType tplTypeCon)) $
+              ETmLam (choiceArg, choiceArgType) $
+                EUpdate $ UExercise (qualify tplTypeCon) choice (EVar chcSelfBinder) Nothing (EVar choiceArg)
+          "_toAnyChoice" ->
+            -- TODO: envLfVersion env `supports` featureAnyType
```
Comment (robin-da) [inline compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs]: Can  we not add the implementation already for the lf versions that support `Any`?

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs [hunk 13]
```diff
@@ -392,7 +392,7 @@ convertGenericTemplate env x
                                             (ETmLam (mkVar "any", typeConAppToType anyContractKeyTy) $ EFromAny keyType $ ERecProj anyContractKeyTy anyContractKeyField $ EVar $ mkVar "any"))
                                   else EBuiltin BEError `ETyApp`
                                        TForall (mkTypeVar "proxy", KArrow KStar KStar) (TApp (TVar $ mkTypeVar "proxy") polyType :-> typeConAppToType anyContractKeyTy :-> TOptional keyType) `ETmApp`
-                                       EBuiltin (BEText "toAnyChoice is not supported in this DAML-LF version")
+                                       EBuiltin (BEText "fromAnyContractKey is not supported in this DAML-LF version")
```
Comment (robin-da) [inline compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs]: Thanks for these!

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs [hunk 14]
```diff
@@ -138,6 +137,8 @@ generateTemplateInstanceModule env externPkgId
           modName <>
           " as X"
         , "import \"" <> packageName <> "\" " <> modName
+        , "import qualified DA.Internal.LF"
+        , "import qualified DA.Internal.Prelude"
```
Comment (robin-da) [inline compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: You will have to import these now as `Sdk.DA.Internal.Prelude` and expose the modules with this alias in `generateAndInstallInstancesPackage` in Packaging.hs.

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs [hunk 15]
```diff
@@ -247,31 +257,36 @@ generateTemplateInstance env typeCon typeParams externPkgId =
...
         LF.mkTApps
-            (LF.TCon (LF.Qualified LF.PRSelf moduleName0 dataTypeCon))
-            (map (LF.TVar . fst) dataParams)
-    templateMethodNames =
+            (LF.TCon (LF.Qualified LF.PRSelf moduleName0 typeCon))
+            (map (LF.TVar . fst) typeParams)
+    methodMapping =
+        map (\funName -> (funName, mkExternalString funName)) methodNames
+    mkExternalString :: T.Text -> String
+    mkExternalString funName =
```
Comment (robin-da) [inline compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: we have copies of `lfTemplateType`, `mkExternalString` and `methodMapping` in `generateChoiceInstance` and `generateTemplateInstance`. We should factor these out.

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs [hunk 16]
```diff
@@ -283,86 +298,153 @@ generateTemplateInstance env typeCon typeParams externPkgId =
...
-            , fun_tick = []
+            , cid_binds = listToBag $ map classMethodStub methodMapping
+            , cid_sigs = []
+            , cid_tyfam_insts = []
+            , cid_datafam_insts = []
+            , cid_overlap_mode = Nothing
             }
+  where
+    app :: LHsType GhcPs -> LHsType GhcPs -> LHsType GhcPs
+    app t1 t2 = noLoc $ HsAppTy noExt t1 t2
```
Comment (robin-da) [inline compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: this is already present in ghc-lib as `mkHsAppTy`.

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs [hunk 17]
```diff
@@ -283,86 +298,153 @@ generateTemplateInstance env typeCon typeParams externPkgId =
...
+
+    body :: LHsType GhcPs =
+      choiceClass `app` arg1 `app` arg2 `app` arg3
+
+    choiceClass :: LHsType GhcPs =
+        noLoc $
+        HsTyVar noExt NotPromoted $
+        noLoc $
+        mkRdrQual (mkModuleName "DA.Internal.Template") $
```
Comment (robin-da) [inline compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: You'll also need to qualify this as `Sdk.DA.Internal.Template`.

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs [hunk 18]
```diff
@@ -283,86 +298,153 @@ generateTemplateInstance env typeCon typeParams externPkgId =
...
+    arg1 :: LHsType GhcPs =
+      noLoc $ convType env lfTemplateType
+
+    arg2 :: LHsType GhcPs =
+      noLoc $ convType env lfChoiceType
+
+    arg3 :: LHsType GhcPs =
+      noLoc $ convType env lfChoiceReturnType
+
+    moduleNameStr = T.unpack $ LF.moduleNameString $ LF.moduleName $ envMod env
```
Comment (robin-da) [inline compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: you can reuse moduleName0 here.

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs [hunk 19]
```diff
@@ -283,86 +298,153 @@ generateTemplateInstance env typeCon typeParams externPkgId =
...
+    moduleName0 = LF.moduleName $ envMod env
+
+    lfTemplateType :: LF.Type =
+        LF.mkTApps
+          (LF.TCon (LF.Qualified LF.PRSelf moduleName0 dataTypeCon))
+          (map (LF.TVar . fst) dataParams)
+
+    tycon :: LF.TypeConName = LF.tplTypeCon template
```
Comment (robin-da) [inline compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: I changed this to match with the style above.

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs [hunk 20]
```diff
@@ -283,86 +298,153 @@ generateTemplateInstance env typeCon typeParams externPkgId =
...
+    templateDT = case NM.lookup tycon (LF.moduleDataTypes (envMod env)) of
+      Just x -> x
+      Nothing -> error $ "Internal error: Could not find template definition for: " <> show tycon
+
+    LF.DefDataType{dataTypeCon,dataParams} = templateDT
+    LF.TemplateChoice{chcArgBinder=(_,lfChoiceType),chcName,chcReturnType=lfChoiceReturnType} = choice
```
Comment (robin-da) [inline compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: I allowed myself  to reformat this, I found it hard to read.

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs (hunk 10 above)
Comment (nickchapman-da) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs]: ok

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs (hunk 11 above)
Comment (nickchapman-da) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs]: yes

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs (hunk 13 above)
Comment (nickchapman-da) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs]: you're welcome!

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs (hunk 15 above)
Comment (nickchapman-da) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: done, except for `mkExternalString` which varies

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs (hunk 16 above)
Comment (nickchapman-da) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: changed to use `mkHsAppTy

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs (hunk 18 above)
Comment (nickchapman-da) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: ok

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs (hunk 14 above)
Comment (nickchapman-da) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: I dont understand what is special about the 2 imports I added. 
Many cases of `DA.Internal.`... without an `Sdk.` prefix were already present in this file.

Please feel free to make the change you think is necessary in this PR.

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs (hunk 17 above)
Comment (nickchapman-da) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: same as above

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs (hunk 12 above)
Comment (nickchapman-da) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs]: As discussed, let's leave this to a following PR

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs (hunk 14 above)
Comment (nickchapman-da) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs]: Now I understand. This change is necessary after rebasing on master.
